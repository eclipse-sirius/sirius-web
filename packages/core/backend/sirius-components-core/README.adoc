= sirius-components-core

== Goal

This project is used to contain the API to manipulate core concepts that are not specific to the business of our application.

== Dependencies

- sirius-components-annotations

This project should not depend on anything else without a good reason to do so.