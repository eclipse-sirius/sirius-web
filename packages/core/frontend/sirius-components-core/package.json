{
  "name": "@eclipse-sirius/sirius-components-core",
  "version": "2024.5.0",
  "author": "Eclipse Sirius",
  "license": "EPL-2.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/eclipse-sirius/sirius-web"
  },
  "publishConfig": {
    "registry": "https://npm.pkg.github.com/"
  },
  "main": "./dist/sirius-components-core.umd.js",
  "module": "./dist/sirius-components-core.es.js",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "exports": {
    ".": {
      "require": "./dist/sirius-components-core.umd.js",
      "import": "./dist/sirius-components-core.es.js"
    }
  },
  "scripts": {
    "start": "vite build --mode 'development' && tsc",
    "build": "vite build && tsc",
    "format": "prettier --write \"src/**/*.{js,ts,tsx,css}\"",
    "format-lint": "prettier --list-different \"src/**/*.{js,ts,tsx,css}\"",
    "test": "vitest --run --config vite.config.js",
    "coverage": "vitest run --config vite.config.js --coverage",
    "publish:local": "yalc push"
  },
  "peerDependencies": {
    "@apollo/client": "3.10.4",
    "@material-ui/core": "4.12.4",
    "@material-ui/icons": "4.11.3",
    "@xstate/react": "1.6.3",
    "graphql": "16.8.1",
    "react": "17.0.2",
    "xstate": "4.32.1"
  },
  "devDependencies": {
    "@apollo/client": "3.10.4",
    "@eclipse-sirius/sirius-components-tsconfig": "*",
    "@material-ui/core": "4.12.4",
    "@material-ui/icons": "4.11.3",
    "@types/react": "17.0.37",
    "@vitejs/plugin-react": "4.0.4",
    "@xstate/react": "1.6.3",
    "@vitest/coverage-v8": "0.34.2",
    "jsdom": "16.7.0",
    "graphql": "16.8.1",
    "prettier": "2.7.1",
    "react": "17.0.2",
    "rollup-plugin-peer-deps-external": "2.2.4",
    "xstate": "4.32.1",
    "typescript": "5.1.6",
    "vite": "4.4.9",
    "vitest": "0.34.2"
  }
}
