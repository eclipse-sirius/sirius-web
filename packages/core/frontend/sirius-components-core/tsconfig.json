{
  "$schema": "https://json.schemastore.org/tsconfig",
  "extends": "@eclipse-sirius/sirius-components-tsconfig/react-library.json",
  "include": ["."],
  "exclude": ["dist", "build", "node_modules"]
}
