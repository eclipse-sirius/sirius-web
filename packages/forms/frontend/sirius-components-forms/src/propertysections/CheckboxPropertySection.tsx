/*******************************************************************************
 * Copyright (c) 2019, 2024 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
import { gql, useMutation } from '@apollo/client';
import { getCSSColor, useMultiToast } from '@eclipse-sirius/sirius-components-core';
import Checkbox from '@material-ui/core/Checkbox';
import FormControl from '@material-ui/core/FormControl';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import FormHelperText from '@material-ui/core/FormHelperText';
import { Theme, makeStyles } from '@material-ui/core/styles';
import { useEffect } from 'react';
import { PropertySectionComponent, PropertySectionComponentProps } from '../form/Form.types';
import { GQLCheckbox } from '../form/FormEventFragments.types';
import {
  CheckboxStyleProps,
  GQLEditCheckboxInput,
  GQLEditCheckboxMutationData,
  GQLEditCheckboxMutationVariables,
  GQLEditCheckboxPayload,
  GQLErrorPayload,
  GQLSuccessPayload,
} from './CheckboxPropertySection.types';
import { PropertySectionLabel } from './PropertySectionLabel';

const useStyle = makeStyles<Theme, CheckboxStyleProps>((theme) => ({
  formControl: {
    alignItems: 'flex-start',
  },
  style: {
    color: ({ color }) => (color ? getCSSColor(color, theme) : theme.palette.primary.light),
    paddingTop: theme.spacing(0.5),
    paddingBottom: theme.spacing(0.5),
  },
  disabled: {},
}));

export const editCheckboxMutation = gql`
  mutation editCheckbox($input: EditCheckboxInput!) {
    editCheckbox(input: $input) {
      __typename
      ... on ErrorPayload {
        messages {
          body
          level
        }
      }
      ... on SuccessPayload {
        messages {
          body
          level
        }
      }
    }
  }
`;

const isErrorPayload = (payload: GQLEditCheckboxPayload): payload is GQLErrorPayload =>
  payload.__typename === 'ErrorPayload';
const isSuccessPayload = (payload: GQLEditCheckboxPayload): payload is GQLSuccessPayload =>
  payload.__typename === 'SuccessPayload';

export const CheckboxPropertySection: PropertySectionComponent<GQLCheckbox> = ({
  editingContextId,
  formId,
  widget,
  readOnly,
}: PropertySectionComponentProps<GQLCheckbox>) => {
  const props: CheckboxStyleProps = {
    color: widget.style?.color ?? null,
  };
  const classes = useStyle(props);

  const [editCheckbox, { loading, error, data }] = useMutation<GQLEditCheckboxMutationData>(editCheckboxMutation);
  const onChange = (event) => {
    const newValue = event.target.checked;
    const input: GQLEditCheckboxInput = {
      id: crypto.randomUUID(),
      editingContextId,
      representationId: formId,
      checkboxId: widget.id,
      newValue,
    };
    const variables: GQLEditCheckboxMutationVariables = {
      input,
    };
    editCheckbox({ variables });
  };

  const { addErrorMessage, addMessages } = useMultiToast();

  useEffect(() => {
    if (!loading) {
      if (error) {
        addErrorMessage('An unexpected error has occurred, please refresh the page');
      }
      if (data) {
        const { editCheckbox } = data;
        if (isErrorPayload(editCheckbox) || isSuccessPayload(editCheckbox)) {
          addMessages(editCheckbox.messages);
        }
      }
    }
  }, [loading, error, data]);

  return (
    <FormControl classes={{ root: classes.formControl }} error={widget.diagnostics.length > 0}>
      <FormControlLabel
        labelPlacement={widget.style?.labelPlacement ?? 'end'}
        label={<PropertySectionLabel editingContextId={editingContextId} formId={formId} widget={widget} />}
        control={
          <Checkbox
            name={widget.label}
            color="default"
            checked={widget.booleanValue}
            onChange={onChange}
            data-testid={widget.label}
            disabled={readOnly || widget.readOnly}
            classes={{ root: classes.style, disabled: classes.disabled }}
          />
        }
      />
      <FormHelperText>{widget.diagnostics[0]?.message}</FormHelperText>
    </FormControl>
  );
};
