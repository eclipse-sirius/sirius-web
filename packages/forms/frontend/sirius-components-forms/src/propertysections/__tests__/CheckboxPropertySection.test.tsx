/*******************************************************************************
 * Copyright (c) 2022, 2024 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { MessageOptions, ToastContext, ToastContextValue } from '@eclipse-sirius/sirius-components-core';
import { act, cleanup, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { afterEach, expect, test, vi } from 'vitest';
import { GQLCheckbox } from '../../form/FormEventFragments.types';
import { CheckboxPropertySection, editCheckboxMutation } from '../CheckboxPropertySection';
import {
  GQLEditCheckboxMutationData,
  GQLEditCheckboxMutationVariables,
  GQLErrorPayload,
  GQLSuccessPayload,
} from '../CheckboxPropertySection.types';

crypto.randomUUID = vi.fn(() => '48be95fc-3422-45d3-b1f9-d590e847e9e1');

afterEach(() => cleanup());

const defaultCheckbox = {
  __typename: 'Checkbox',
  id: 'checkboxId',
  label: 'CheckboxLabel',
  iconURL: [],
  hasHelpText: false,
  diagnostics: [],
  booleanValue: false,
  readOnly: false,
};

const checkboxWithStyle: GQLCheckbox = {
  ...defaultCheckbox,
  style: {
    color: '#de1000',
    labelPlacement: 'top',
  },
};

const checkboxWithEmptyStyle: GQLCheckbox = {
  ...defaultCheckbox,
  style: {
    color: '',
    labelPlacement: 'end',
  },
};

const readOnlyCheckbox: GQLCheckbox = {
  ...checkboxWithEmptyStyle,
  readOnly: true,
};

const editCheckboxVariables: GQLEditCheckboxMutationVariables = {
  input: {
    id: '48be95fc-3422-45d3-b1f9-d590e847e9e1',
    editingContextId: 'editingContextId',
    representationId: 'formId',
    checkboxId: 'checkboxId',
    newValue: true,
  },
};
const successPayload: GQLSuccessPayload = { messages: [], __typename: 'SuccessPayload' };
const editCheckboxSuccessData: GQLEditCheckboxMutationData = { editCheckbox: successPayload };

const editCheckboxErrorPayload: GQLErrorPayload = {
  __typename: 'ErrorPayload',
  messages: [
    {
      body: 'An error has occurred, please refresh the page',
      level: 'ERROR',
    },
    {
      body: 'FeedbackMessage',
      level: 'INFO',
    },
  ],
};
const editCheckboxErrorData: GQLEditCheckboxMutationData = {
  editCheckbox: editCheckboxErrorPayload,
};

const mockEnqueue = vi.fn<[string, MessageOptions?], void>();

const toastContextMock: ToastContextValue = {
  enqueueSnackbar: mockEnqueue,
};

test('should render the checkbox', () => {
  const { container } = render(
    <MockedProvider>
      <ToastContext.Provider value={toastContextMock}>
        <CheckboxPropertySection
          editingContextId="editingContextId"
          formId="formId"
          widget={checkboxWithStyle}
          readOnly={false}
        />
      </ToastContext.Provider>
    </MockedProvider>
  );
  expect(container).toMatchSnapshot();
});

test('should render a readOnly checkbox', () => {
  const { container } = render(
    <MockedProvider>
      <ToastContext.Provider value={toastContextMock}>
        <CheckboxPropertySection
          editingContextId="editingContextId"
          formId="formId"
          widget={readOnlyCheckbox}
          readOnly
        />
      </ToastContext.Provider>
    </MockedProvider>
  );
  expect(container).toMatchSnapshot();
});

test('should send mutation when clicked', async () => {
  let editCheckboxCalled = false;
  const editCheckboxSuccessMock: MockedResponse<Record<string, any>> = {
    request: {
      query: editCheckboxMutation,
      variables: editCheckboxVariables,
    },
    result: () => {
      editCheckboxCalled = true;
      return { data: editCheckboxSuccessData };
    },
  };

  const mocks = [editCheckboxSuccessMock];
  const { container } = render(
    <MockedProvider mocks={mocks}>
      <ToastContext.Provider value={toastContextMock}>
        <CheckboxPropertySection
          editingContextId="editingContextId"
          formId="formId"
          widget={checkboxWithStyle}
          readOnly={false}
        />
      </ToastContext.Provider>
    </MockedProvider>
  );
  expect(container).toMatchSnapshot();

  const element: HTMLInputElement | null = screen.getByTestId('CheckboxLabel').querySelector('input');
  expect(element && element.checked).toBe(false);

  await act(async () => {
    if (element) {
      userEvent.click(element);
      await new Promise((resolve) => setTimeout(resolve, 0));

      await waitFor(() => {
        expect(editCheckboxCalled).toBeTruthy();
        expect(container).toMatchSnapshot();
      });
    }
  });
});

test('should display the error received', async () => {
  let editCheckboxCalled = false;
  const editCheckboxErrorMock: MockedResponse<Record<string, any>> = {
    request: {
      query: editCheckboxMutation,
      variables: editCheckboxVariables,
    },
    result: () => {
      editCheckboxCalled = true;
      return { data: editCheckboxErrorData };
    },
  };

  const mocks = [editCheckboxErrorMock];
  const { baseElement } = render(
    <MockedProvider mocks={mocks}>
      <ToastContext.Provider value={toastContextMock}>
        <CheckboxPropertySection
          editingContextId="editingContextId"
          formId="formId"
          widget={checkboxWithStyle}
          readOnly={false}
        />
      </ToastContext.Provider>
    </MockedProvider>
  );
  expect(baseElement).toMatchSnapshot();

  const element: HTMLInputElement | null = screen.getByTestId('CheckboxLabel').querySelector('input');
  expect(element && element.checked).toBe(false);

  await act(async () => {
    if (element) {
      userEvent.click(element);
      await new Promise((resolve) => setTimeout(resolve, 0));

      await waitFor(() => {
        expect(editCheckboxCalled).toBeTruthy();
        expect(mockEnqueue).toHaveBeenCalledTimes(2);
        expect(baseElement).toMatchSnapshot();
      });
    }
  });
});

test('should render the checkbox without style', () => {
  const { baseElement } = render(
    <MockedProvider>
      <ToastContext.Provider value={toastContextMock}>
        <CheckboxPropertySection
          editingContextId="editingContextId"
          formId="formId"
          widget={checkboxWithEmptyStyle}
          readOnly={false}
        />
      </ToastContext.Provider>
    </MockedProvider>
  );
  expect(baseElement).toMatchSnapshot();
});

test('should render the checkbox with style', () => {
  const { baseElement } = render(
    <MockedProvider>
      <ToastContext.Provider value={toastContextMock}>
        <CheckboxPropertySection
          editingContextId="editingContextId"
          formId="formId"
          widget={checkboxWithStyle}
          readOnly={false}
        />
      </ToastContext.Provider>
    </MockedProvider>
  );
  expect(baseElement).toMatchSnapshot();
});

test('should render the checkbox with empty style', async () => {
  const { baseElement } = render(
    <MockedProvider>
      <ToastContext.Provider value={toastContextMock}>
        <CheckboxPropertySection
          editingContextId="editingContextId"
          formId="formId"
          widget={checkboxWithEmptyStyle}
          readOnly={false}
        />
      </ToastContext.Provider>
    </MockedProvider>
  );
  expect(baseElement).toMatchSnapshot();
});

test('should render a checkbox with help hint', () => {
  const { container } = render(
    <MockedProvider>
      <ToastContext.Provider value={toastContextMock}>
        <CheckboxPropertySection
          editingContextId="editingContextId"
          formId="formId"
          widget={{ ...checkboxWithEmptyStyle, hasHelpText: true }}
          readOnly
        />
      </ToastContext.Provider>
    </MockedProvider>
  );
  expect(container).toMatchSnapshot();
});

test('should render a readOnly checkbox from widget properties', () => {
  const { container } = render(
    <MockedProvider>
      <ToastContext.Provider value={toastContextMock}>
        <CheckboxPropertySection
          editingContextId="editingContextId"
          formId="formId"
          widget={readOnlyCheckbox}
          readOnly={false}
        />
      </ToastContext.Provider>
    </MockedProvider>
  );
  expect(container).toMatchSnapshot();
});
