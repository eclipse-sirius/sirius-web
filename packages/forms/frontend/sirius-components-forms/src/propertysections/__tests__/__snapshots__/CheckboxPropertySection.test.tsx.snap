// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should display the error received 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root makeStyles-formControl-37"
    >
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementTop"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-41 MuiCheckbox-root makeStyles-style-38 makeStyles-style-40"
          data-testid="CheckboxLabel"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-44"
              data-indeterminate="false"
              name="CheckboxLabel"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          <div
            class="makeStyles-propertySectionLabel-45"
          >
            <h6
              class="MuiTypography-root makeStyles-typography-48 MuiTypography-subtitle2"
            >
              CheckboxLabel
            </h6>
          </div>
        </span>
      </label>
      <p
        class="MuiFormHelperText-root"
      />
    </div>
  </div>
</body>
`;

exports[`should display the error received 2`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root makeStyles-formControl-37"
    >
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementTop"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-41 MuiCheckbox-root makeStyles-style-38 makeStyles-style-40"
          data-testid="CheckboxLabel"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-44"
              data-indeterminate="false"
              name="CheckboxLabel"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          <div
            class="makeStyles-propertySectionLabel-45"
          >
            <h6
              class="MuiTypography-root makeStyles-typography-48 MuiTypography-subtitle2"
            >
              CheckboxLabel
            </h6>
          </div>
        </span>
      </label>
      <p
        class="MuiFormHelperText-root Mui-focused"
      />
    </div>
  </div>
</body>
`;

exports[`should render a checkbox with help hint 1`] = `
<div>
  <div
    class="MuiFormControl-root makeStyles-formControl-85"
  >
    <label
      class="MuiFormControlLabel-root Mui-disabled"
    >
      <span
        aria-disabled="true"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-89 MuiCheckbox-root makeStyles-style-86 makeStyles-style-88 PrivateSwitchBase-disabled-91 Mui-disabled makeStyles-disabled-87 Mui-disabled Mui-disabled"
        data-testid="CheckboxLabel"
        tabindex="-1"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-92"
            data-indeterminate="false"
            disabled=""
            name="CheckboxLabel"
            type="checkbox"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
      </span>
      <span
        class="MuiTypography-root MuiFormControlLabel-label Mui-disabled MuiTypography-body1"
      >
        <div
          class="makeStyles-propertySectionLabel-93"
        >
          <h6
            class="MuiTypography-root makeStyles-typography-96 MuiTypography-subtitle2"
          >
            CheckboxLabel
          </h6>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorSecondary"
            focusable="false"
            style="margin-left: 8px; font-size: 16px;"
            title=""
            viewBox="0 0 24 24"
          >
            <path
              d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
            />
          </svg>
        </div>
      </span>
    </label>
    <p
      class="MuiFormHelperText-root"
    />
  </div>
</div>
`;

exports[`should render a readOnly checkbox 1`] = `
<div>
  <div
    class="MuiFormControl-root makeStyles-formControl-13"
  >
    <label
      class="MuiFormControlLabel-root Mui-disabled"
    >
      <span
        aria-disabled="true"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-17 MuiCheckbox-root makeStyles-style-14 makeStyles-style-16 PrivateSwitchBase-disabled-19 Mui-disabled makeStyles-disabled-15 Mui-disabled Mui-disabled"
        data-testid="CheckboxLabel"
        tabindex="-1"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-20"
            data-indeterminate="false"
            disabled=""
            name="CheckboxLabel"
            type="checkbox"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
      </span>
      <span
        class="MuiTypography-root MuiFormControlLabel-label Mui-disabled MuiTypography-body1"
      >
        <div
          class="makeStyles-propertySectionLabel-21"
        >
          <h6
            class="MuiTypography-root makeStyles-typography-24 MuiTypography-subtitle2"
          >
            CheckboxLabel
          </h6>
        </div>
      </span>
    </label>
    <p
      class="MuiFormHelperText-root"
    />
  </div>
</div>
`;

exports[`should render a readOnly checkbox from widget properties 1`] = `
<div>
  <div
    class="MuiFormControl-root makeStyles-formControl-97"
  >
    <label
      class="MuiFormControlLabel-root Mui-disabled"
    >
      <span
        aria-disabled="true"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-101 MuiCheckbox-root makeStyles-style-98 makeStyles-style-100 PrivateSwitchBase-disabled-103 Mui-disabled makeStyles-disabled-99 Mui-disabled Mui-disabled"
        data-testid="CheckboxLabel"
        tabindex="-1"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-104"
            data-indeterminate="false"
            disabled=""
            name="CheckboxLabel"
            type="checkbox"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
      </span>
      <span
        class="MuiTypography-root MuiFormControlLabel-label Mui-disabled MuiTypography-body1"
      >
        <div
          class="makeStyles-propertySectionLabel-105"
        >
          <h6
            class="MuiTypography-root makeStyles-typography-108 MuiTypography-subtitle2"
          >
            CheckboxLabel
          </h6>
        </div>
      </span>
    </label>
    <p
      class="MuiFormHelperText-root"
    />
  </div>
</div>
`;

exports[`should render the checkbox 1`] = `
<div>
  <div
    class="MuiFormControl-root makeStyles-formControl-1"
  >
    <label
      class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementTop"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiCheckbox-root makeStyles-style-2 makeStyles-style-4"
        data-testid="CheckboxLabel"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-8"
            data-indeterminate="false"
            name="CheckboxLabel"
            type="checkbox"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
      >
        <div
          class="makeStyles-propertySectionLabel-9"
        >
          <h6
            class="MuiTypography-root makeStyles-typography-12 MuiTypography-subtitle2"
          >
            CheckboxLabel
          </h6>
        </div>
      </span>
    </label>
    <p
      class="MuiFormHelperText-root"
    />
  </div>
</div>
`;

exports[`should render the checkbox with empty style 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root makeStyles-formControl-73"
    >
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-77 MuiCheckbox-root makeStyles-style-74 makeStyles-style-76"
          data-testid="CheckboxLabel"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-80"
              data-indeterminate="false"
              name="CheckboxLabel"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          <div
            class="makeStyles-propertySectionLabel-81"
          >
            <h6
              class="MuiTypography-root makeStyles-typography-84 MuiTypography-subtitle2"
            >
              CheckboxLabel
            </h6>
          </div>
        </span>
      </label>
      <p
        class="MuiFormHelperText-root"
      />
    </div>
  </div>
</body>
`;

exports[`should render the checkbox with style 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root makeStyles-formControl-61"
    >
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementTop"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-65 MuiCheckbox-root makeStyles-style-62 makeStyles-style-64"
          data-testid="CheckboxLabel"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-68"
              data-indeterminate="false"
              name="CheckboxLabel"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          <div
            class="makeStyles-propertySectionLabel-69"
          >
            <h6
              class="MuiTypography-root makeStyles-typography-72 MuiTypography-subtitle2"
            >
              CheckboxLabel
            </h6>
          </div>
        </span>
      </label>
      <p
        class="MuiFormHelperText-root"
      />
    </div>
  </div>
</body>
`;

exports[`should render the checkbox without style 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root makeStyles-formControl-49"
    >
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-53 MuiCheckbox-root makeStyles-style-50 makeStyles-style-52"
          data-testid="CheckboxLabel"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-56"
              data-indeterminate="false"
              name="CheckboxLabel"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          <div
            class="makeStyles-propertySectionLabel-57"
          >
            <h6
              class="MuiTypography-root makeStyles-typography-60 MuiTypography-subtitle2"
            >
              CheckboxLabel
            </h6>
          </div>
        </span>
      </label>
      <p
        class="MuiFormHelperText-root"
      />
    </div>
  </div>
</body>
`;

exports[`should send mutation when clicked 1`] = `
<div>
  <div
    class="MuiFormControl-root makeStyles-formControl-25"
  >
    <label
      class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementTop"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-29 MuiCheckbox-root makeStyles-style-26 makeStyles-style-28"
        data-testid="CheckboxLabel"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-32"
            data-indeterminate="false"
            name="CheckboxLabel"
            type="checkbox"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
      >
        <div
          class="makeStyles-propertySectionLabel-33"
        >
          <h6
            class="MuiTypography-root makeStyles-typography-36 MuiTypography-subtitle2"
          >
            CheckboxLabel
          </h6>
        </div>
      </span>
    </label>
    <p
      class="MuiFormHelperText-root"
    />
  </div>
</div>
`;

exports[`should send mutation when clicked 2`] = `
<div>
  <div
    class="MuiFormControl-root makeStyles-formControl-25"
  >
    <label
      class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementTop"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-29 MuiCheckbox-root makeStyles-style-26 makeStyles-style-28"
        data-testid="CheckboxLabel"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-32"
            data-indeterminate="false"
            name="CheckboxLabel"
            type="checkbox"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
      >
        <div
          class="makeStyles-propertySectionLabel-33"
        >
          <h6
            class="MuiTypography-root makeStyles-typography-36 MuiTypography-subtitle2"
          >
            CheckboxLabel
          </h6>
        </div>
      </span>
    </label>
    <p
      class="MuiFormHelperText-root Mui-focused"
    />
  </div>
</div>
`;
